### PR TITLE
Permission host mismatch

### PR DESCRIPTION
**Pull request by jumde**
Fixes: https://github.com/brave/browser-laptop/issues/13471
Depends: https://github.com/brave/muon/pull/540

## Test plan
1. Open https://google.com
2. Change the address in the URL bar to sms://ggle.com. Verify the external permission message is for  sms://ggle.com and not https://google.com
3. Launch a webserver with test.html:
`<a href="sms://ggle.com">Test</a>`
4. Open http(s)://<webserver>/test.html
5. Click on the hyperlink, verify the permission is requested for `sms://ggle.com` and not `<webserver>`

### CODE REVIEW

File: app/filtering.js [hunk 1]
```diff
@@ -495,9 +496,10 @@ function registerPermissionHandler (session, partition) {
         }
       }
 
+      let parsedMainFrameUrl = new URL(mainFrameUrl)
```
Comment (diracdeltas) [inline app/filtering.js]: can this use `urlParse`?

**Comment by diracdeltas** [comment]
could you please add a test plan with the expected behavior?

File: app/filtering.js (hunk 1 above)
Comment (jumde) [reply app/filtering.js]: With `urlParse`:

```
tps://oogle.com/search?q=s%3A%2F%2Fwww.lcebook.com%2F
{ hash: '',
  hostname: '',
  host: '',
  href: 'tps://oogle.com/search?q=s%3A%2F%2Fwww.lcebook.com%2F',
  path: '//oogle.com/search?q=s%3A%2F%2Fwww.lcebook.com%2F',
  pathname: '//oogle.com/search',
  port: '',
  protocol: 'tps:',
  query: 'q=s%3A%2F%2Fwww.lcebook.com%2F',
  search: '?q=s%3A%2F%2Fwww.lcebook.com%2F',
  origin: '' }
```

Comment (jumde) [reply app/filtering.js]: with `new URL`
```
URL {
  href: tps://oogle.com/search?q=s%3A%2F%2Fwww.lcebook.com%2F
  protocol: tps:
  hostname: oogle.com
  pathname: /search
  search: ?q=s%3A%2F%2Fwww.lcebook.com%2F
}
```

Comment (jumde) [reply app/filtering.js]: Fails here: `<a href="tp://ttt.com">Test Url</a>`

File: app/common/urlParse.js [hunk 2]
```diff
@@ -5,15 +5,7 @@
...
-// muon.url.parse is not available in all environments (ex: unittests)
-let urlParse
-try {
-  urlParse = muon.url.parse
-} catch (e) {
-  // TODO: move to the new node URL API: https://nodejs.org/api/url.html#url_url
-  urlParse = require('url').parse
-}
-
+let urlParse = require('url').parse
```
Comment (diracdeltas) [inline app/common/urlParse.js]: there are some security issues with Node's default URL parser - can this behavior be fixed in Muon's url parser?

Comment (jumde) [reply app/common/urlParse.js]: yes, working on that.

File: app/filtering.js [hunk 3]
```diff
@@ -511,7 +511,7 @@ function registerPermissionHandler (session, partition) {
             {text: locale.translation('deny')},
             {text: locale.translation('allow')}
           ],
-          frameOrigin: getOrigin(mainFrameUrl),
+          frameOrigin: getOrigin(origin),
```
Comment (diracdeltas) [inline app/filtering.js]: i think you want mainFrameUrl here

Comment (jumde) [reply app/filtering.js]: The variable names  need to be updated here. 

In https://github.com/brave/muon/pull/540. I updated the `origin` to be the security origin of the current frame, and the `mainFrameUrl` to the URL requesting the permission.

**Comment by diracdeltas** [comment]
@jumde previously, the `openExternal` permission (as I understood it), was granted to the main frame origin, not the origin that is being opened. Ex: if you go to brave.com and click the `mailto:` link at the bottom of the page, the expected behavior is that it prompts you whether to allow brave.com to open external URLs. 

If we are going to change it to the behavior specified in your test plan, we need to also:
1. update the permission message to "Allow URL_TO_BE_OPENED to be opened in an external application?" instead of "Allow MAIN_FRAME_URL to open external applications"?
2. delete the existing openExternal permission settings, since those are no longer valid with this change

i agree that the behavior in this PR makes more sense than the existing behavior, but i also want to note that the scope of these changes is larger than https://github.com/brave/browser-laptop/issues/13471 (which can be fixed by just changing the displayed URL to be the external URL in the case where a main frame URL doesn't exist).

File: js/lib/urlutil.js [hunk 4]
```diff
@@ -470,7 +470,10 @@ const UrlUtil = {
       if (parsed.protocol === 'about:') {
         return [parsed.protocol, parsed.path].join('')
       }
-      return parsed.origin.replace(/\/+$/, '')
+      if (parsed.origin.length !== 0) {
+        return parsed.origin.replace(/\/+$/, '')
+      }
+      return parsed.href.replace(/\/+$/, '')
```
Comment (diracdeltas) [inline js/lib/urlutil.js]: is this change still needed? i don't think this works on arbitrary URLs, ex:

```
muon.url.parse('sms://bing/kfdk').href.replace(/\/+$/, '')  // returns "sms://bing/kfdk"
```

File: app/filtering.js [hunk 5]
```diff
@@ -427,35 +427,35 @@ function registerPermissionHandler (session, partition) {
...
+      mainFrameOrigin = getOrigin(mainFrameOrigin)
+      settings = siteSettings.getSiteSettingsForURL(appState.get('siteSettings'), mainFrameOrigin)
+      tempSettings = siteSettings.getSiteSettingsForURL(appState.get('temporarySiteSettings'), mainFrameOrigin)
     }
 
     let response = []
 
-    if (origin == null) {
+    if (requestingUrl == null) {
       response = new Array(permissionTypes.length)
```
Comment (bsclifton) [inline app/filtering.js]: This looks to be the first real change- everything else above this line is just renaming, is that correct?

Comment (jumde) [reply app/filtering.js]: Yes!

File: app/filtering.js [hunk 6]
```diff
@@ -497,7 +497,7 @@ function registerPermissionHandler (session, partition) {
 
       // Display 'Brave Browser' if the origin is null; ex: when a mailto: link
       // is opened in a new tab via right-click
-      const message = locale.translation('permissionMessage').replace(/{{\s*host\s*}}/, origin || 'Brave Browser').replace(/{{\s*permission\s*}}/, permissions[permission].action)
+      const message = locale.translation('permissionMessage').replace(/{{\s*host\s*}}/, getOrigin(requestingUrl) || 'Brave Browser').replace(/{{\s*permission\s*}}/, permissions[permission].action)
```
Comment (bsclifton) [inline app/filtering.js]: Do you have an example of where the `setPermissionRequestHandler` handler gets called from? I see the definition in Muon, but am not sure when the handler is called?

Comment (bsclifton) [reply app/filtering.js]: Looking at the code in Muon...
`mainFrameOrigin` is the `current_origin`
`requestingUrl` maps to `requesting_origin`

Super glad you fixed this mismatch issue 😄

File: app/filtering.js [hunk 7]
```diff
@@ -530,7 +530,7 @@ function registerPermissionHandler (session, partition) {
           response[index] = result
           if (persist) {
             // remember site setting for this host
-            appActions.changeSiteSetting(origin, permission + 'Permission', result, isPrivate)
+            appActions.changeSiteSetting(requestingUrl, permission + 'Permission', result, isPrivate)
```
Comment (diracdeltas) [inline app/filtering.js]: I'm confused why we are getting the permissions for the main frame origin in https://github.com/brave/browser-laptop/pull/13475/files#diff-8a9bac13ef6264feb2b6da7c18d86b3cR452, but here we apply permissions by `requestingUrl` instead of `mainFrameOrigin`. seems like these two lines should either both use `requestingUrl` or both use `mainFrameOrigin`

File: app/filtering.js [hunk 8]
```diff
@@ -427,35 +427,35 @@ function registerPermissionHandler (session, partition) {
...
+      mainFrameOrigin = isPDFOrigin ? 'PDF Viewer' : 'Brave Browser'
       // display on all tabs
-      mainFrameUrl = null
+      requestingUrl = null
       // Lookup by exact host pattern match since 'Brave Browser' is not
       // a parseable URL
-      settings = siteSettings.getSiteSettingsForHostPattern(appState.get('siteSettings'), origin)
-      tempSettings = siteSettings.getSiteSettingsForHostPattern(appState.get('temporarySiteSettings'), origin)
-    } else if (mainFrameUrl.startsWith('magnet:')) {
+      settings = siteSettings.getSiteSettingsForHostPattern(appState.get('siteSettings'), mainFrameOrigin)
```
Comment (diracdeltas) [inline app/filtering.js]: are you sure this and the other `getSiteSettingsForHostPattern` lines aren't supposed to use  `getOrigin(requestingUrl)` since that's what is actually shown in the message?

Comment (diracdeltas) [reply app/filtering.js]: https://github.com/brave/browser-laptop/pull/13475/files#diff-8a9bac13ef6264feb2b6da7c18d86b3cR500

File: app/filtering.js (hunk 7 above)
Comment (jumde) [reply app/filtering.js]: Good catch! fixed

File: app/filtering.js [hunk 9]
```diff
@@ -427,35 +427,35 @@ function registerPermissionHandler (session, partition) {
...
       // display on all tabs
-      mainFrameUrl = null
+      mainFrameOrigin = null
       // Lookup by exact host pattern match since 'Brave Browser' is not
       // a parseable URL
-      settings = siteSettings.getSiteSettingsForHostPattern(appState.get('siteSettings'), origin)
-      tempSettings = siteSettings.getSiteSettingsForHostPattern(appState.get('temporarySiteSettings'), origin)
-    } else if (mainFrameUrl.startsWith('magnet:')) {
+      settings = siteSettings.getSiteSettingsForHostPattern(appState.get('siteSettings'), requestingUrl)
+      tempSettings = siteSettings.getSiteSettingsForHostPattern(appState.get('temporarySiteSettings'), requestingUrl)
```
Comment (diracdeltas) [inline app/filtering.js]: you have to do `requestingUrl = getOrigin(requestingUrl)` before calling these settings since these were previously saved by origin not full URL

Comment (diracdeltas) [reply app/filtering.js]: or better yet, define `const requestingOrigin = getOrigin(requestingUrl)` in the function scope so that it's harder to confuse URL with origin

File: app/filtering.js [hunk 10]
```diff
@@ -427,35 +427,35 @@ function registerPermissionHandler (session, partition) {
...
       // This covers an edge case where you open a magnet link tab, then disable Torrent Viewer
       // and restart Brave. I don't think it needs localization. See 'Brave Browser' above.
-      origin = 'Magnet URL'
+      requestingUrl = 'Magnet URL'
     } else {
       // Strip trailing slash
-      origin = getOrigin(origin)
-      settings = siteSettings.getSiteSettingsForURL(appState.get('siteSettings'), origin)
-      tempSettings = siteSettings.getSiteSettingsForURL(appState.get('temporarySiteSettings'), origin)
+      requestingUrl = getOrigin(requestingUrl)
```
Comment (diracdeltas) [inline app/filtering.js]: see https://github.com/brave/browser-laptop/pull/13475/files#r177257591

File: app/filtering.js (hunk 6 above)
Comment (diracdeltas) [inline app/filtering.js]: can also be simplified using https://github.com/brave/browser-laptop/pull/13475/files#r177257591

File: app/filtering.js (hunk 7 above)
Comment (diracdeltas) [reply app/filtering.js]: see https://github.com/brave/browser-laptop/pull/13475/files#r177257591 - this should apply the change by requesting origin, not full URL

File: js/lib/urlutil.js (hunk 4 above)
Comment (diracdeltas) [reply js/lib/urlutil.js]: let's remove this diff if it's no longer needed

Comment (jumde) [reply js/lib/urlutil.js]: This is still needed, `muon.url.parse` for `muon.url.parse('sms://bing/kfdk')` returns NULL origin. The warning message if this change is not present, the warning message will read `Allow Brave Browser to open an external application?`

File: app/filtering.js [hunk 11]
```diff
@@ -427,35 +427,35 @@ function registerPermissionHandler (session, partition) {
...
-    const isBraveOrigin = origin.startsWith(`chrome-extension://${config.braveExtensionId}/`)
-    const isPDFOrigin = origin.startsWith(`${pdfjsOrigin}/`)
+    const isBraveOrigin = mainFrameOrigin.startsWith(`chrome-extension://${config.braveExtensionId}/`)
+    const isPDFOrigin = mainFrameOrigin.startsWith(`${pdfjsOrigin}/`)
     let settings
     let tempSettings
-    if (mainFrameUrl === appUrlUtil.getBraveExtIndexHTML() || isPDFOrigin || isBraveOrigin) {
+    let requestingOrigin = getOrigin(requestingUrl)
+
+    if (requestingOrigin === appUrlUtil.getBraveExtIndexHTML() || isPDFOrigin || isBraveOrigin) {
```
Comment (diracdeltas) [inline app/filtering.js]: you want `requestingUrl` in this line i think

File: js/lib/urlutil.js (hunk 4 above)
Comment (diracdeltas) [reply js/lib/urlutil.js]: @jumde what if instead of this change, you just change https://github.com/brave/browser-laptop/pull/13475/files#diff-7b5b041c814805dd8496488bc6a53d3cR469 to `parsed.origin != undefined` so that the null origin will trigger https://github.com/brave/browser-laptop/pull/13475/files#diff-7b5b041c814805dd8496488bc6a53d3cR479 ?

also please add a test case in the getOrigin unit tests

Comment (jumde) [reply js/lib/urlutil.js]: Host is also null for arbitrary URLs:

```
Url: hps://nodejs.org/en/download/
{ hash: '',
  hostname: '',
  host: '',
  href: 'hps://nodejs.org/en/download/',
  path: '//nodejs.org/en/download/',
  pathname: '//nodejs.org/en/download/',
  port: '',
  protocol: 'hps:',
  query: '',
  search: '',
  origin: '' }```

Comment (diracdeltas) [reply js/lib/urlutil.js]: @jumde i see. what if instead of changing the behavior of `getOrigin`, we just change https://github.com/brave/browser-laptop/pull/13475/files#diff-8a9bac13ef6264feb2b6da7c18d86b3cR500 such to `requestingOrigin || requestingUrl || 'Brave Browser'` so it displays correctly?

i am reluctant to change getOrigin since it's used in various places for security checks

**Comment by diracdeltas** [comment]
@jumde thanks for making the changes. PR looks good to me now :). i will test it once the muon build is out.

File: app/filtering.js [hunk 12]
```diff
@@ -471,7 +471,7 @@ function registerPermissionHandler (session, partition) {
         response.push(false)
       } else if (permission === 'fullscreen' &&
         // The Torrent Viewer extension is always allowed to show fullscreen media
-        origin.startsWith('chrome-extension://' + config.torrentExtensionId)) {
+        mainFrameOrigin.startsWith('chrome-extension://' + config.torrentExtensionId)) {
```
Comment (diracdeltas) [inline app/filtering.js]: this throws an error when you go to http://www.orimi.com/pdf-test.pdf and click the fullscreen button

should be `requestingOrigin` instead of mainFrameOrigin i think

Comment (diracdeltas) [reply app/filtering.js]: actually it shouldn't - https://github.com/brave/browser-laptop/pull/13475/files#diff-8a9bac13ef6264feb2b6da7c18d86b3cR440 set the mainFrameOrigin to null deliberately

so i think this just has to check if mainFrameOrigin is null

File: app/filtering.js [hunk 13]
```diff
@@ -427,35 +427,35 @@ function registerPermissionHandler (session, partition) {
...
-      mainFrameUrl = null
+      mainFrameOrigin = null
       // Lookup by exact host pattern match since 'Brave Browser' is not
       // a parseable URL
-      settings = siteSettings.getSiteSettingsForHostPattern(appState.get('siteSettings'), origin)
-      tempSettings = siteSettings.getSiteSettingsForHostPattern(appState.get('temporarySiteSettings'), origin)
-    } else if (mainFrameUrl.startsWith('magnet:')) {
+      settings = siteSettings.getSiteSettingsForHostPattern(appState.get('siteSettings'), requestingOrigin)
+      tempSettings = siteSettings.getSiteSettingsForHostPattern(appState.get('temporarySiteSettings'), requestingOrigin)
+    } else if (requestingOrigin.startsWith('magnet:')) {
```
Comment (diracdeltas) [inline app/filtering.js]: this `if` block` is actually not necessary anymore since it will show the magnet URL itself

**Comment by bsclifton** [comment]
I am an idiot and just merged again without squashing 😛

**Comment by bsclifton** [comment]
master 569450262ec1ee43dac7bdb8b3c73606fd8e9f15
0.23.x ac2b5fdc1ad6990030d2c466160037a560d01069
0.22.x cd7a8781d6fb6d0307865276de9e2ebac4fc91f0
